posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,172 @@
+
+```lean
+example (p : Nat → Prop) : ∃ x : Nat, p x ∧ x = 5 := by
+  simp only [existsAndEq]
+  -- Remaining goal: `⊢ p 5`
+
+example (p q : Nat → Prop) : ∃ x : Nat, p x ∧ x = 5 ∧ q x := by
+  simp only [existsAndEq]
+  -- Remaining goal: `⊢ p 5 ∧ q 5`
+```
+
+Intuitively, one can think of `existsAndEq` as a "parametric" lemma whose right hand side is allowed to vary shape depending on the left hand side.
+
+Roughly speaking, the way this metaprogram operates is a follows: whenever an expression of the form `∃ a, P a` with `P a = Q ∧ R`  is encountered:
+- Recursively traverse the expression of the predicate inside the existential quantifier to try and detect an equality `a = a'` by splitting any `∧` that is found along the way into its components.
+- If an equality is found, construct a proof that `P a` implies `a = a'`.
+- Construct a proof that `(∃ a, P a) = P a'` using the following theorem
+```lean
+theorem exists_of_imp_eq {α : Sort u} {p : α → Prop} (a : α) (h : ∀ b, p b → a = b) :
+    (∃ b, p b) = p a
+```
+
+### Performance optimisation: The `reduceIte` simproc
+
+The `reduceIte` simproc is designed to take expressions of the form `if P then a else b` (aka `ite P a b`) and replace them with `a` or `b`, depending on whether `P` simplify to `True` or `False`.
+
+```lean
+example : (if 37 * 0 then 1 else 2) = 1 := by
+  -- Works since `simp` can simplify `37 * 0` to `True`
+  -- because it knows the lemma `mul_zero a : a * 0 = 0`.
+  simp only [↓reduceIte]
+
+example (X : Type) (P : Prop) (a b : X) : (if P ∨ ¬ P then a else b) = a := by
+  -- Works since `simp` can simplify `P ∨ ¬ P` to `True`.
+  simp only [↓reduceIte]
+
+open Classical
+
+example : (if FermatLastTheorem then 1 else 2) = 1 := by
+  --This fails because `simp` can't simplify `FermatLastTheorem` to `True` or `False`
+  simp only [↓reduceIte]
+  -- Remaining goal: `⊢ (if FermatLastTheorem then 1 else 2) = 1`
+  sorry -- See https://imperialcollegelondon.github.io/FLT for how to solve this `sorry` ;)
+```
+
+The down arrow `↓` in `simp only [↓reduceIte]` indicates that `reduceIte` is being called as a *preprocedure*.
+This means that `reduceIte` is run *before* the expressions `a` and `b` are simplified.
+Concretely, `simp` needs only simplify one of the expressions `a` and `b`, as the other one is discarded upfront.
+For example, `↓reduceIte` allows `simp` to avoid touching `bigComplicatedExpression` at all in the following:
+```
+example : (if 37 * 0 = 0 then 0 else bigComplicatedExpression) = 0 := by
+  simp only [↓reduceIte]
+```
+
+Notice that one could always replace `simp [↓reduceIte]` with `simp [ite_cond_eq_true, ite_cond_eq_false]`, but the latter would simplify both branches of the `ite`, instead of merely the one that is actually relevant.
```

Comment:
I misunderstood this at first read. What about:
```suggestion
Notice that one could always replace `simp [↓reduceIte]` with `simp [ite_cond_eq_true, ite_cond_eq_false]` which would leave the same final goal state, but the would simplify inside both branches of the `ite`, instead of simplifying away the irrelevant one before passing to subexpressions.
```